Add support for building just a subset of runtime tests

trylek opened the pull request:
This change adds four new options to the runtime test build command:

1) test <test project path relative to src\tests> - build just that
one test.

2) dir <directory relative to src\tests> - build all tests in the
directory.

3) tree <directory relative to src\tests> - build all tests in the
given subtree.

4) all - use clean rebuild (i.e. don't apply incrementalism)
when building the tests.

Thanks

Tomas

/cc @dotnet/runtime-infrastructure

trylek committed:
Add support for building just a subset of runtime tests

This change adds four new options to the runtime test build command:

1) test <test project path relative to src\tests> - build just that
one test.

2) dir <directory relative to src\tests> - build all tests in the
directory.

3) tree <directory relative to src\tests> - build all tests in the
given subtree.

4) all - use clean rebuild (i.e. don't apply incrementalism)
when building the tests.

Thanks

Tomas

am11 commented on src/tests/build.sh:
```diff
@@ -541,6 +575,10 @@ __DistroRid=""
 __DoCrossgen2=0
 __CompositeBuildMode=0
 __TestBuildMode=
+__BuildTestProject=""
+__BuildTestDir=""
+__BuildTestTree=""
```
nit: (like elsewhere in this script) empty string is not needed.

trylek replied:
Changed to "%3B" to overcome MSBuild parameter shenanigans.

am11 replied:
Heh, we also had to do similar thing once in https://github.com/dotnet/runtime/blob/03601a7a817068adb67639673db2010eaaf32d2a/eng/build.sh#L471-L473

jkoritzinsky commented:
What's the expected behavior if multiple of these new commands are supplied? Does one override the others, do multiple take effect, or does the script error out?

naricc commented:
> What's the expected behavior if multiple of these new commands are supplied? Does one override the others, do multiple take effect, or does the script error out?

It looks like multiple seperate commands take a effect (-dir/-tree), but using for example -test twice for two different tests doesn't. I'm not sure this is wrong exactly, but it is a little odd?

trylek commented:
Right now only the last command of each type is honored; I can trivially change them to item groups by using the semicolon separator and honoring all instances of each; I'll update the PR after our team meeting.

trylek committed:
Allow multiple specifications of 'test', 'dir' and 'tree'

trylek committed:
Fix Unix version of the script

AntonLapounov commented on src/tests/build.sh:
```diff
@@ -432,6 +439,11 @@ usage_list+=("-buildtestwrappersonly: only build the test wrappers.")
+usage_list+=("-test:xxx - only build a single test project");
+usage_list+=("-dir:xxx - build all tests in a given directory");
+usage_list+=("-tree:xxx - build all tests in a given subtree");
+usage_list+=("-all:xxx - rebuild all the specified tests");
```
`-all` has no argument:
```suggestion
usage_list+=("-all - rebuild all the specified tests");
```

AntonLapounov commented:
> all - use clean rebuild (i.e. don't apply incrementalism)

Would it not be better to name this option `-clean` then?

trylek committed:
Fix copy & paste bug in help string per Anton's PR feedback

trylek commented:
> Would it not be better to name this option -clean then?

@AntonLapounov - I don't know; I'm not too happy with <code>all</code> either but I've thought about this for quite a while and I haven't found any other ideal name:

1) <code>src\tests\build clean</code> looks like a request to just clean up the test builds, not to rebuild everything.
2) <code>src\tests\build rebuild</code> is what I originally had but to my dismay I found out that <code>src\tests\build.sh</code> already implements this option by deleting all test artifacts prior to the build.

I'll be happy for suggestions.

naricc commented:
> > Would it not be better to name this option -clean then?
> 
> @AntonLapounov - I don't know; I'm not too happy with `all` either but I've thought about this for quite a while and I haven't found any other ideal name:
> 
> 1. `src\tests\build clean` looks like a request to just clean up the test builds, not to rebuild everything.
> 2. `src\tests\build rebuild` is what I originally had but to my dismay I found out that `src\tests\build.sh` already implements this option by deleting all test artifacts prior to the build.
> 
> I'll be happy for suggestions.

Maybe -cleanbuild ?

trylek commented:
I think you're right, that sounds reasonable to me.

trylek committed:
Remove the 'all' option and implement 'rebuild' in cmd per Anton's PR feedback

It also turns out that the rebuild cleanup was in the wrong place
in the build.sh script - it was done as part of generate_layout i.e.
after the native and managed test components have been built i.o.w.
using it ended up building the tests and deleting them afterwards.

Thanks

Tomas

trylek committed:
Replace the 'all' option with 'rebuild' in build.cmd help text

trylek committed:
Fix overlooked $ sign in build.sh

AntonLapounov reviewed (approved):
Looks good, thank you!

fanyang-mono reviewed (changes requested):
I am so happy to see this change. No more hacking for this. Please update the corresponding doc for this new capability.

trylek committed:
Describe the new options in test workflow docs per Fan's PR feedback

trylek commented:
@fanyang-mono - I have added description of the new options to the workflow docs per your suggestion. Can you please take another look when you have a chance and let me know if you find that sufficient or possibly suggest additional doc updates?

fanyang-mono commented on docs/workflow/testing/coreclr/unix-test-instructions.md:
```diff
@@ -40,6 +40,35 @@ The `src/tests/build.sh` script generates the Core_Root folder, which contains t
+lengthy (especially in `-priority1` mode) and unnecessary when working on a particular test.
+
+1) `-test:<test-project>` - build a particular test project specified by its project file path,
+either absolute or relative to `src/tests`. The option can be specified multiple times on the command
```
Please mention that the option that they could provide multiple items and separate them by semicolon.

trylek replied:
Updated in 9th commit.

fanyang-mono commented on docs/workflow/testing/coreclr/unix-test-instructions.md:
```diff
@@ -40,6 +40,35 @@ The `src/tests/build.sh` script generates the Core_Root folder, which contains t
+**Example**: `src/tests/build.sh -test:JIT/Methodical/divrem/div/i4div_cs_do.csproj`
+
+2) `-dir:<test-folder>` - build all test projects within a given directory path, either absolute
+or relative to `src/tests`. The option can be specified multiple times on the command line to request
```
same here.

trylek replied:
Updated in 9th commit.

fanyang-mono commented on docs/workflow/testing/coreclr/unix-test-instructions.md:
```diff
@@ -40,6 +40,35 @@ The `src/tests/build.sh` script generates the Core_Root folder, which contains t
+**Example**: `src/tests/build.sh -dir:JIT/Methodical/Arrays/huge`
+
+3) `-tree:<root-folder>` - build all test projects within the subtree specified by its root path,
+either absolute or relative to `src/tests`. The option can be specified multiple times on the command
```
same here.

trylek replied:
Updated in 9th commit.

trylek committed:
Further expand documentation per Fan's PR feedback

am11 commented:
Seems like it is not compatible with all projects under src/tests.
e.g. ILVerification.Tests.csproj

```sh
$ git clean -xdf
$ ./build.sh -c release
$ src/tests/build.sh -release -test:ilverify/ILVerification.Tests.csproj 
...
...
Managed tests build success!
Creating test wrappers...
Microsoft (R) Build Engine version 17.0.0-preview-21411-04+6ca861613 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

/Users/am11/projects/runtime/.dotnet/sdk/6.0.100-rc.1.21411.28/MSBuild.dll -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/Users/am11/projects/runtime/.dotnet/sdk/6.0.100-rc.1.21411.28/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/Users/am11/projects/runtime/.dotnet/sdk/6.0.100-rc.1.21411.28/dotnet.dll -maxcpucount -verbosity:m /bl:/Users/am11/projects/runtime/artifacts/log/Release/build_test_wrappers_coreclr.binlog /consoleloggerparameters:Summary /fileloggerparameters1:WarningsOnly;LogFile=/Users/am11/projects/runtime/artifacts/log/Check_Test_Build.OSX.x64.Release.wrn /fileloggerparameters2:ErrorsOnly;LogFile=/Users/am11/projects/runtime/artifacts/log/Check_Test_Build.OSX.x64.Release.err /fileloggerparameters:Verbosity=normal;LogFile=/Users/am11/projects/runtime/artifacts/log/Check_Test_Build.OSX.x64.Release.log /nodereuse:false /p:BuildWrappers=true /p:TestBuildMode= /p:TargetsWindows= /p:TargetOS=OSX /p:Configuration=Release /p:TargetArchitecture=x64 /p:RuntimeFlavor=coreclr /Users/am11/projects/runtime/src/tests/run.proj
CSC : warning CS2008: No source files specified. [/Users/am11/projects/runtime/src/tests/Common/test_dependencies/test_dependencies.csproj]
  test_dependencies -> /Users/am11/projects/runtime/artifacts/tests/coreclr/OSX.x64.Release/Common/test_dependencies/test_dependencies/test_dependencies.dll
/Users/am11/projects/runtime/src/tests/run.proj(342,9): error MSB4184: The expression "[MSBuild]::MakeRelative('', /**/*.sh)" cannot be evaluated. The path is empty. (Parameter 'path')

Build FAILED.

CSC : warning CS2008: No source files specified. [/Users/am11/projects/runtime/src/tests/Common/test_dependencies/test_dependencies.csproj]
/Users/am11/projects/runtime/src/tests/run.proj(342,9): error MSB4184: The expression "[MSBuild]::MakeRelative('', /**/*.sh)" cannot be evaluated. The path is empty. (Parameter 'path')
    1 Warning(s)
    1 Error(s)
```
Without `-test:...` argument ilverification test project builds fine and we can run it with:
```sh
bash artifacts/tests/coreclr/OSX.x64.Release/ilverify/ILVerification.Tests.sh \
    -coreroot=$(pwd)/artifacts/tests/coreclr/OSX.x64.Release/Tests/Core_Root
```

am11 commented on docs/workflow/testing/coreclr/unix-test-instructions.md:
```diff
@@ -40,6 +40,38 @@ The `src/tests/build.sh` script generates the Core_Root folder, which contains t
+line to request building several individual projects; alternatively, a single occurrence of the option
+can represent several project paths separated by semicolons.
+
+**Example**: `src/tests/build.sh -test:JIT/Methodical/divrem/div/i4div_cs_do.csproj;JIT/Methodical/divrem/div/i8div_cs_do.csproj`
```
In all Unix-y shells including powershell, semicolon is a statement terminator. This example only builds i4div_cs_do.csproj. I think having just one approach of multiple `-test:` occurrences was enough really. Now we support both approaches to achieve the same thing, with different workwounds each.

trylek commented:
@am11 - for the ILVerification.Tests.csproj, sorry about that; as I described in the accompanying documentation, the new options are currently orthogonal to the priority value; as you can easily see the test ILVerification.Tests is Pri1,

https://github.com/dotnet/runtime/blob/26b20d0ba3dc9553173372dc0dda03b1fba91258/src/tests/ilverify/ILVerification.Tests.csproj#L5

and so it won't build unless you specify -priority1 in the src/tests/build.sh command line. I understand it's somewhat countra-intuitive but the other option - forcibly setting priority to 1 whenever a -test  / -dir / -tree option is specified it not ideal either because it creates even more complex internal logic where seemingly identical things work differently - e.g. "-tree:." would end up building something different than when -tree is not specified and / or -tree / -dir containing a single test project would behave differently than specifying -test with the same project name.

My hope is that, based on my proposal

https://github.com/dotnet/runtime/issues/54512

we should be able to reduce the overall number of CoreCLR tests and ultimately get rid of the Pri0 / Pri1 distinction; I plan to start preparatory cleanup work in this direction during the oncoming quality week.

For the other comment regarding inappropriate use of semicolon in Unix, perhaps we can just change the semicolon to colon on Unix akin to how search paths are normally stored.
